Random space structures

unid15 opened the pull request:
![](http://puu.sh/nwQWv/cd255d3474.png)
- At the beginning of each round, a random number of pre-mapped structures (later - vaults) is placed around the map
- Exact amount is rand(1,(number of possible vaults)), maximum amount is 10 (unless somebody adds more areas)
- Vaults are stored in maps/randomVaults/ as .dmm files
- The vaults are only placed in designated areas around the map
  ![](http://puu.sh/nwR5Q/60357d2319.png)
- Each area can only house one vault inside
- This PR includes 5 high quality vaults:  [stash in an asteroid (guns=toys)](http://puu.sh/nwZbx/0bff8a9aa9.png), [ice cream truck crash](http://puu.sh/nwZdG/f1712a4673.png), [hivebot factory](http://puu.sh/nAB8C/101072fa8f.png), [clown base](http://puu.sh/nBQ2I/6c98716711.png), [ruined RUST generator by @PJB3005](https://camo.githubusercontent.com/ad226e04acbfb21427a160c065b5b1bd81b50659/68747470733a2f2f646c2e64726f70626f7875736572636f6e74656e742e636f6d2f752f3130373032383032362f5368617265582f323031362f30332f647265616d6d616b65725f323031362d30332d31315f32312d34332d31352e706e67)
- Ice cream truck crash has 3 temp guns (guarded by 6 polar bears)
- Asteroid stash has some chest mimics and a bat helmet
- Hivebot factory has three superglue bottles, a fireaxe, an assault rifle and a syndicate flaregun. The assault rifle is guarded by tons of hivebots, superglue is guarded by two cyber horrors. Fireaxe and syndicate flaregun are hidden in a tiny r-wall room with a cyberhorror.
- Clown base has a bunch of powerful banana peels, a buttbot, H.O.N.K weapons and a genuine clown PDA hidden

**STATUS:**
- works and 100% tested
- vaults may still need some work

Shadowmech88 commented:
:+1:

PJB3005 commented:
OH SHIT SON :+1:

DrCelt commented:
Aha! I'd this idea for #6789 and it's good this'll exist for that. Perhaps consider using datums for the sorting process, though? Otherwise, :+1: :ok_hand:

Shadowmech88 commented:
Hypothetically, how difficult would it be to procedurally generate the structures themselves? Or have a single vault that's the designated "random" vault, and just procedurally generate that specific one if the map spawns it?

DrCelt commented:
@Shadowmech88 Not difficult if you want it to be randomly generated. Very difficult if you don't want that random generation to not look like complete ass

Shadowmech88 commented:
What about doing that thing that Spelunky does where you pre-make chunks, and then have the engine slap some of the chunks together in a random order?

PJB3005 commented:
Same issue @Shadowmech88 because this isn't Spelunky.

Skullyton commented:
Is there any way we can make random "vaults" for this?

Got ideas for a few large ones, small ones, etc.

icantthinkofanameritenow commented:
@Skullyton 
I think maybe it could work if we borrowed the stuff used for generating the hidden rooms on the mining asteroid, but I'm not entirely sure. 

I support this idea- it'll give people a good reason to go and explore space beyond looking for the clown roid.

unid15 commented:
Randomly generated vaults are gonna work on a different code than this, that's for sure! And it's gonna take a lot of work.
Anyways, I just *100%* tested this and it seems to be working fine. Currently these are the 3 vaults that can spawn (spoiler alert they're shit):

[doomed soviet satelite](http://puu.sh/nwZ8O/957da2bf2c.png), [stash in an asteroid (guns=toys)](http://puu.sh/nwZbx/0bff8a9aa9.png), [ice cream truck crash](http://puu.sh/nwZdG/f1712a4673.png)

9600bauds commented:
Dude this is sweet as FUCK, but please for the love of God spread out the spawners over the many z-levels we have

Skullyton commented:
@unid15 Nice, how do we make those tho?

unid15 commented:
They're already spread out over all 5 z-levels

9600bauds commented:
Oh alright, I saw the screenshot of all of them spawning in the same z-level and couldn't gleam from the code if it was that way or not

unid15 commented:
@Skullyton they're .dmm map files, so they can be made in Dream Maker (just like normal maps)

icantthinkofanameritenow commented:
@unid15 

They're decent proofs of concept at least. The full ones should probably have some more appealing loot, though. Maybe the relatively weaker adminbus items that more or less function like recolors of existing gear, like the T-51b power armor.

CptWad commented:
SWEET
How easy is it to add more possible things to the list of possible spawners?

unid15 commented:
You just have to put the file into a folder and add its filename to the list in code/modules/randomMaps/vaults.dm

CptWad commented:
noice

clusterfack commented:
Does this move the station, or is the station usually not centered? Because its kind of triggering me.

CptWad commented:
Thats derelict and its been like that since the space quadrupling

CptWad commented:
Wheres restraunt.dmm?

SkowronX commented:
Derelict plasmaman vessels with hostile plasmamen when

Exxion commented:
Holy SHIT

FudgePucker commented:
The ice cream truck needs some better loot if its gonna be guarded by six bears

CptWad commented:
How bout a temp gun? for the ice cream truck?

Probe1 commented:
:+1:

PJB3005 commented on code/controllers/master_controller.dm:
```diff
@@ -225,6 +225,15 @@ datum/controller/game_controller/proc/cachedamageicons()
 		log_startup_progress("Not generating minimaps - SKIP_MINIMAP_GENERATION found in config/config.txt")
 
+	if(!config.skip_vault_generation)
+		spawn()
```
Due to this `spawn()` the `Finished initializations` message will come up before this even starts to generate the structures.

PJB3005 commented on code/modules/randomMaps/vaults.dm:
```diff
@@ -0,0 +1,98 @@
+//If you want more vaults, feel free to add more subtypes of /area/random_vault. You don't have to add these subtypes to any lists or anything - just map it and the game will handle the rest.
+
+//"/area/random_vault" DOESN'T spawn any vaults!!!
+/area/random_vault/v1
```
Fucking areas.

THANKS BYOND.

To be honest this is probably the rare exception where I'm fine with map instances, but whatever.

zashkuar commented:
roid stash is fine unlike others
Also, look into leaked goon code, they have some nice random stuff, like small abadoned factory or ship with servers and robots

Shadowmech88 commented:
Should we move the station to the center of its Z-level now that the Z-level's been expanded?

unid15 commented:
What are you guys talking about ![](http://puu.sh/nydWf/9eb4156c15.png) it's in the middle

icantthinkofanameritenow commented:
@CptWad

Have you already forgotten that temp guns are useless after Celt's hypothermia update? Even when emagged they're only good for killing slimes. Not to mention that something that can be mass produced with protolathes hardly makes up for fighting through six bears to get to it. 

Remember, the riskier it is to get to the loot the more rewarding it ought to be.

9600bauds commented:
A pre-emagged temp gun then

icantthinkofanameritenow commented:
@zashkuar

Not really. It's just a Batman helmet (not even the full suit, at that), and a syndicate balloon along with some space cash. That's some lame loot if you ask me. 

If anything the satellite is a better example, since it has access to items impossible to find anywhere else outside of special circumstances (the crossbow and the Soviet vending machine).

9600bauds commented:
I'm going to stress this once again:
Randomized loot spawners are easy to implement and it would be a very good time to start using them more

icantthinkofanameritenow commented:
@9600bauds
Even when emagged the temp gun is little more than a joke. It would be better to just replace it with something else that would justify the difficulty of getting to it. Either that, or remove the bears entirely since the rewards for killing them simply aren't worth it now. After all, you need to provide motivation to look for these vaults long after the thrill of exploration and discovery wears off- and the only way to do that is via the lure of loot.

Now that I think of it, what about all the guns that Deity added a while back that were never available outside of Summon Guns and adminbus? I'm sure a Mosin-Nagant wouldn't be out of place on the Soviet satellite.

DrCelt commented:
>Have you already forgotten that temp guns are useless after Celt's hypothermia update? 

Please stop parroting this myth. They're no more or less effective than they were pre hypothermia. That said, they'd be a shite loot item because they are awful anyway.

Arthurdentist commented:
instead of arguing about temp guns lets try and map some new space content to go with this. Here is my shit attempt which I threw together 10 minutes after reading this PR. 

![devils bar](https://cloud.githubusercontent.com/assets/6353372/13588715/e267fe6c-e4c9-11e5-833d-80ee4aa7649d.png)

Jeroen52 commented:
What about an abandoned space arcade with a very large chance (or 100%) for an Outbomb Cuban Pete machine?

CptWad commented:
shit I was gonna do a few once it was merged so it was less of a hassle but I could whip some shit up
also whats the size limit for the things, not that ill push it but I just need to know

9600bauds commented:
Make the dice have a 10% chance to be an E20

clusterfack commented:
Ho ho thats devious

Jeroen52 commented:
@9600bauds Maybe a derelict library?

ihadtoregisterforthis commented:
How does this affect meteors?

icantthinkofanameritenow commented:
@9600bauds 
Or better yet, the cursed dice.

CptWad commented:
@ihadtoregisterforthis
if its on the station z-level it would sorta but on other z-levels it wont do shit

9600bauds commented:
A few problems I've found so far
 
 - When the game starts, it says "skeleton" and a picture of a newcop. I don't know what that's supposed to be.
 - Vaults can spawn in the station's z-level, is this intended?
 - Well... The vaults show up in the minimaps if you clear your cache and login again.
![dreamseeker_2016-03-08_14-54-44](https://cloud.githubusercontent.com/assets/6526157/13610724/e4958ae6-e53d-11e5-9762-4f4c61c2d427.png)

unid15 commented:
Yeah spawning on z-level 1 is intended. Will fix the other two

unid15 commented:
>When the game starts, it says "skeleton" and a picture of a newcop. I don't know what that's supposed to be.

oh god this makes zero sense

when you become an observer (only from the round-start screen), an invisible skeleton spawns in the fucking nullspace and shits a runtime whenever ANY sound is played

but if you spawn normally as a human, everything works perfectly and :100:% flawlessly

send help

Shadowmech88 commented:
>when you become an observer (only from the round-start screen), an invisible skeleton spawns in the fucking nullspace and shits a runtime whenever ANY sound is played

Literally what.

unid15 commented:
Literally this

![](http://puu.sh/nztOy/81c7d1ae47.png)

![](http://puu.sh/nztPR/64fd9f7625.png)

![](http://puu.sh/nzuvO/54d6b29582.png)

PJB3005 commented:
What the fuck? 

If that isn't a BYOND issue I don't know what is....

FudgePucker commented:
'Spookie skeleton in nullspace' should be an arcade game that spawns only in a vault. Its attack will be runtimes.

unid15 commented:
I deleted the doomed satelite from my hard drive, removed all references to it in the code, sprayed my PC with holy water and suddenly everything started working again

If you guys have it saved on your computer for some reason, delete it now before it kills your family and turns you into a nullspace skeleton

CptWad commented:
jesus christ how terrifying
Was there something on the satellite that caused all this or is it just a mystery

Zth-- commented:
Get a copy so we can make a creepypasta

unid15 commented:
Fixed minimaps for good, fixed lighting, added abandoned hivebot factory with good rewards (assault rifle, 3x superglue, fireaxe, flare gun) - the conveyor belts don't work

![](http://puu.sh/nAB8C/101072fa8f.png)

The doomed satelite still exists in the vaults folder

9600bauds commented:
How are those emitters powered

unid15 commented:
Vaults are always powered, like shuttles, so the answer is *magic*

CptWad commented:
The Hivebot on the outside is going to wander off and end up on the station 50% of the time

icantthinkofanameritenow commented:
Maybe replace the assault rifle with something more exotic? In an average round RnD makes a few dozen assault rifles at the bare minimum. What about a silenced pistol or something instead?

Arthurdentist commented:
Could you generate ores in the 'asteroid' part of that vault with good chances for decent ores? Diamond et al?

Skullyton commented:
That's a specific asteroid wall type, 'Good yield'.

The metaclub asteroid's made of the stuff, hence why you rarely see gibtonite on it

ihadtoregisterforthis commented:
Would it be possible to have categories of structures so a very dangerous room doesn't threaten the non-explorers?

Vermilingua commented:
I think we're going to need to expand the spacepod bay, shit's gonna get busy in there after this.

Skullyton commented:
Spacepods would need to be better than a glorified jetpack first.

Main two things added to them that would make them perfect would be a contained storage for stashing loot, and some way of locking it beyond taking the cell out when you go wandering

PJB3005 commented:
Space pods you can walk inside when?

unid15 commented:
You can pick up anchored items, you just can't pull them out of the way or put them in lockers

PJB3005 commented on code/modules/randomMaps/vaults.dm:
```diff
@@ -0,0 +1,102 @@
+				for(var/turf/new_turf in L)
+					new_turf.flags |= NO_MINIMAP //f u c k minimaps
+
+				message_admins("<span class='info'>Loaded [path_file]: [formatJumpTo(locate(vault_x, vault_y, vault_z))]!!!")
```
Don't you think 3 exclamation marks is a tad redundant?

Shadowmech88 commented:
>you can pick up anchored items

What property of projectiles prevents them from being picked up, then? Because they can't be picked up. I thought it was the anchoring that did it.

unid15 commented:
@Shadowmech88 they can't be clicked with the mouse[!](unless you alt+click the ground and click on the projectile from the popup menu)

Shadowmech88 commented:
@unid15 
Is there some mechanic that prevents you from clicking them?
What if you were to right-click and select "pick up" from there? Could you pick up a projectile frozen by time-stop, theoretically?

Exxion commented:
I believe you can just make `attack_hand()` not call the parent to make it impossible to pick up

unid15 commented:
shh

icantthinkofanameritenow commented:
@Skullyton 
Perhaps it would be helpful to "borrow" some of the stuff Goonstation did with their spacepods.

DrCelt commented:
@icantthinkofanameritenow You know the problems you had when porting something from TG? Consider that we are far far far closer to TG than goon. It'd be nigh impossible at this point to truly borrow anything, and would be far easier to just make our own.

icantthinkofanameritenow commented:
@DrCelt 
Yeah...that brings back some unpleasant memories. 

Still, it's something we can use for inspiration. If nothing else, it does seem kind of odd that there's only one armor type for space pods that can be manufactured. But that's another issue entirely.

unid15 commented:
@ihadtoregisterforthis what do you mean? The structures only spawn in space, and you're free to avoid them if you're not well equipped enough

PJB3005 commented:
![](https://dl.dropboxusercontent.com/u/107028026/ShareX/2016/03/dreammaker_2016-03-11_20-16-08.png)

I am making a thing!

ihadtoregisterforthis commented:
@unid15 Many simple mobs can walk on space, and if they start on the station z-level they can eventually reach it with ease.
Other use for these categories would be so some exploration is needed for the best stuff instead of, for example, ordering a borg to RCD some floors in space for loot.

In addition smaller rooms catch fewer meteors than large chunks of derelict arks.

unid15 commented:
I'll probably remove vaults from z-1 then

Shadowmech88 commented:
Or remove the ability of simple mobs to fly through space.

PJB3005 commented:
https://github.com/unid15/vgstation13/pull/5

Skullyton commented:
Wait, we can do PRs to this PR?

Oh boy, time for shenanigans

MusgardTiger commented:
FEISTY

NigglyWiggly commented:
I have some ideas for these and I might start throwing some stuff together if I can figure out how to into mapping.

- Derelict Space port 'Argo' with some lower tech stuff inside, flashing neon sign outside the port
- Remains of a Grey space craft, (I finally have an excuse to try to put my various Grey themed sprites in the game)
- Maybe an ancient (modern) Satellite? 
- Various space bars, 'gas' stations, and 'truck' stops

Shadowmech88 commented:
One of the vaults needs to be a nonfunctional ship that can be restored and piloted.
By "ship" I'm thinking more of the constructable shuttles that can be done via adminbus, rather than the salvage shuttle.

unid15 commented:
Removed all vaults from the station's z-level, so meteor waves wouldn't get affected by them. Stopped ranged hivebots from leaving the hivebot factory (by turning the ones outside the doors into stationary hivebot sentries)
Will squash commits when [NFD]

clusterfack commented:
Fix conflicts and squash then its good to go

unid15 commented:
Alright all done

(map conflicts = satan)

Exxion commented:
For the record, we have a .dmm merge engine, so just merging the two conflicting branches locally should fix the conflicts unless an individual tile was changed in both of them
I've never tested that though so it may or may not actually work
It's very old

unid15 commented:
I know but the engine didn't work

Exxion commented:
Oh
Did it say there were conflicts it couldn't fix or did it just not work?

unid15 commented:
I don't remember but I'm pretty sure it tried fixing them and then gave up

Exxion commented:
Oh good
It's probably not completely broken, then

PJB3005 commented:
HAPPENING

Vermilingua commented:
ayyyyy
On 14 Mar 2016 5:36 pm, "PJB3005" <notifications@github.com> wrote:

> HAPPENING
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/d3athrow/vgstation13/pull/8683#issuecomment-196166702>
> .
>
